### PR TITLE
Adds initial features implementation for siddhi - io - nats extension.

### PR DESCRIPTION
**Pull request by Ketharan**
## Purpose
Providing the ability to communicate (publish/subscribe) with nats-streaming server from siddhi application. The source configuration can be used to subscribe and sink configuration  can be used to publish to the nats-streaming server. 

## Goals
Adds specific sink and source configuration which can be used in siddhi applications to initiallize the communication to nats-streaming server.

## Approach
Adding the implementation for the available siddhi IO extension template using the officially supported [nats-streaming](https://github.com/nats-io/java-nats-streaming) java client provided by NATS developers. 

## User stories
Please find the user story document [here](https://docs.google.com/document/d/1vGX6F1AvFR6cY9akO2JJwy1lnnGyBJWBNPiedcL-GfY/edit?usp=sharing).

## Release note
Siddhi IO nats extension helps to communicate with nats streaming server from siddhi application.  

## Documentation
Please find the documentation for source configuration in the annotations of this [file.](https://github.com/Ketharan/siddhi-io-nats/blob/e08fb968c5780b30f52633482fc43660339cd8fe/component/src/main/java/org/wso2/extension/siddhi/io/nats/source/NATSSource.java#L48)

Please find the documentation for sink configuration in the annotations of this [file.](https://github.com/Ketharan/siddhi-io-nats/blob/e08fb968c5780b30f52633482fc43660339cd8fe/component/src/main/java/org/wso2/extension/siddhi/io/nats/sink/NATSSink.java#L47)

## Training
N/A

## Marketing
N/A

## Automation tests
 - Unit tests 
Unit test included. Code coverage (70 %)

## Security checks
 - Followed secure coding standards in http://wso2.com/technical-reports/wso2-secure-engineering-guidelines? yes
 - Ran FindSecurityBugs plugin and verified report? yes
 - Confirmed that this PR doesn't commit any keys, passwords, tokens, usernames, or other secrets? yes

## Samples
This example shows how to subscribe to a NATS subject with all supporting configurations.With the following configuration the source identified as 'nats-client' will subscribes to a subject named as 'SP_NATS_INPUT_TEST' which resides in a nats instance with a cluster id of 'test-cluster', running in localhost and listening to the port 4222 for client connection. This subscription will receive all the messages from 100th in the subject.

@source(type='nats', @map(type='text'), "
                   destination='SP_NATS_INPUT_TEST', 
                   bootstrap.servers='nats://localhost:4222',"
                   client.id='nats_client',"
                   server.id='test-cluster',"
                   queue.group.name = 'group_nats',"
                   durable.name = 'nats-durable',"
                   subscription.sequence = '100')
define stream inputStream (name string, age int, country string);

This example shows how to publish to a NATS subject with all supporting configurations. With the following configuration the sink identified as 'nats-client' will publish to a subject named as 'SP_NATS_OUTPUT_TEST' which resides in a nats instance with a cluster id of 'test-cluster', running in localhost and listening to the port 4222 for client connection.

@sink(type='nats', @map(type='xml')
                   destination='SP_NATS_OUTPUT_TEST', "
                   bootstrap.servers='nats://localhost:4222',"
                   client.id='nats_client',"
                   server.id='test-cluster')
define stream outputStream (name string, age int, country string);

## Test environment
-  JDK version: 1.8
- Operating system: Ubuntu 16.04
- Nats-streaming-server: 0.11.2

## Learning

- [NATS](https://nats.io/) official documentation.
- [Nats - streaming](https://nats.io/documentation/streaming/nats-streaming-intro/) official documentation.
- nats-streaming-server github [readme](https://github.com/nats-io/nats-streaming-server).
- nats-streaming java client [repository](https://github.com/nats-io/java-nats-streaming ).
- nats-streaming [slack](https://natsio.slack.com/join/shared_invite/enQtMzE2NDkxNDI2NTE1LTc5ZDEzYTkwYWZkYWQ5YjY1MzBjMWZmYzA5OGQxMzlkMGQzMjYxNGM3MWYxMjNiYmNjNzIwMTVjMWE2ZDgxZGM) channel.
- [Javadoc](http://javadoc.io/doc/io.nats/java-nats-streaming/2.1.2) for nats-streaming java client.
- [Design and architectural decisions](https://github.com/nats-io/nats-general) of NATS.

### CODE REVIEW

File: component/siddhi-io-nats.iml [hunk 1]
```diff
@@ -0,0 +1,83 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (pcnfernando) [inline component/siddhi-io-nats.iml]: Exclude .iml files

Comment (ksdperera) [inline component/siddhi-io-nats.iml]: We don't need to commit .iml files

Comment (Ketharan) [reply component/siddhi-io-nats.iml]: Resolved with the commit 508fb521854cc962d479d5cb9d2099973b53d4bb

Comment (Ketharan) [reply component/siddhi-io-nats.iml]: Resolved with the commit 508fb521854cc962d479d5cb9d2099973b53d4bb

File: docs/images/siddhi-logo-w.svg [hunk 2]
```diff
@@ -0,0 +1,163 @@
+<?xml version="1.0" encoding="UTF-8" standalone="no"?>
+<!--
+  ~ Copyright (c) 2017, WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
```
Comment (AnuGayan) [inline docs/images/siddhi-logo-w.svg]: Please change the license header to 2018. Check other places as well

File: component/src/test/java/org/wso2/extension/siddhi/io/nats/source/NATSSourceTestCase.java [hunk 3]
```diff
@@ -0,0 +1,892 @@
...
+                + "<country>US</country></event></events>");
+        NATSClient.publish("nats-test11","<events><event><name>RAHEEM</name><age>47</age>"
+                + "<country>GERMANY</country></event></events>");
+        NATSClient.publish("nats-test11","<events><event><name>JANE</name><age>36</age>"
+                + "<country>US</country></event></events>");
+        NATSClient.publish("nats-test11","<events><event><name>LAKE</name><age>19</age>"
+                + "<country>GERMANY</country></event></events>");
+        Thread.sleep(3000);
+
+
```
Comment (AnuGayan) [inline component/src/test/java/org/wso2/extension/siddhi/io/nats/source/NATSSourceTestCase.java]: Remove unnecessary new lines. Check other places as well

File: pom.xml [hunk 4]
```diff
@@ -0,0 +1,155 @@
...
+            </dependency>
+        </dependencies>
+    </dependencyManagement>
+    <build>
+        <plugins>
+            <plugin>
+                <groupId>org.apache.maven.plugins</groupId>
+                <artifactId>maven-release-plugin</artifactId>
+                <configuration>
+                    <preparationGoals>clean install</preparationGoals>
```
Comment (Niveathika92) [inline pom.xml]: Have to call documentation deploy profile

File: README.md [hunk 5]
```diff
@@ -1 +1,44 @@
-# siddhi-io-nats
\ No newline at end of file
+﻿siddhi-io-nats
+======================================
+
```
Comment (Niveathika92) [inline README.md]: Add a basic description. Do a mkdocs build and serve optionally to check all doc are generated correctly

File: component/pom.xml [hunk 6]
```diff
@@ -0,0 +1,145 @@
...
+                <configuration>
+                    <instructions>
+                        <Bundle-SymbolicName>${project.artifactId}</Bundle-SymbolicName>
+                        <Bundle-Name>${project.artifactId}</Bundle-Name>
+                        <Export-Package>
+                            org.wso2.extension.siddhi.io.nats,
+                            org.wso2.extension.siddhi.io.nats.*
+                        </Export-Package>
+                        <Import-Package>
+                            ;resolution:=optional
```
Comment (Niveathika92) [inline component/pom.xml]: Import explicitly

File: component/pom.xml [hunk 7]
```diff
@@ -0,0 +1,145 @@
...
+                        </Import-Package>
+                        <DynamicImport-Package>*</DynamicImport-Package>
+                        <Include-Resource>
+                            META-INF=target/classes/META-INF
```
Comment (Niveathika92) [inline component/pom.xml]: Add maven resources

File: component/src/main/java/org/wso2/extension/siddhi/io/nats/sink/NATSSink.java [hunk 8]
```diff
@@ -0,0 +1,238 @@
...
+    private static final Logger log = Logger.getLogger(NATSSink.class);
+    private StreamingConnection streamingConnection;
+    private OptionHolder optionHolder;
+    private StreamDefinition streamDefinition;
+    private Option destination;
+    private String clusterId;
+    private String clientId;
+    private String natsUrl;
+
+    /**
```
Comment (Niveathika92) [inline component/src/main/java/org/wso2/extension/siddhi/io/nats/sink/NATSSink.java]: Remove inherited code for more readability

File: component/src/test/java/org/wso2/extension/siddhi/io/nats/sink/NATSSinkTestCase.java [hunk 9]
```diff
@@ -0,0 +1,343 @@
...
+        } catch (SiddhiAppValidationException e) {
+            Assert.assertTrue(e.getMessage().contains("Option 'destination' does not exist in the configuration of "
+                    + "'sink:nats'"));
+        }
+    }
+
+    /**
+     * If invalid NATS url provided then {@link SiddhiAppCreationException} will be thrown
+     */
+    @Test
```
Comment (Niveathika92) [inline component/src/test/java/org/wso2/extension/siddhi/io/nats/sink/NATSSinkTestCase.java]: Assert using @Test(expectedException=.......)

File: mkdocs.yml [hunk 10]
```diff
@@ -0,0 +1,16 @@
+site_name: Siddhi-IO-Nats
```
Comment (Niveathika92) [inline mkdocs.yml]: Update as per mkdocs site uprade

File: docs/images/siddhi-logo-w.svg (hunk 2 above)
Comment (Ketharan) [reply docs/images/siddhi-logo-w.svg]: Resolved with the commit 7e05e862337f827854551f7cdea54c88c72623e4.

File: component/src/test/java/org/wso2/extension/siddhi/io/nats/source/NATSSourceTestCase.java (hunk 3 above)
Comment (Ketharan) [reply component/src/test/java/org/wso2/extension/siddhi/io/nats/source/NATSSourceTestCase.java]: Resolved with the commit 7e05e862337f827854551f7cdea54c88c72623e4.

File: pom.xml (hunk 4 above)
Comment (Ketharan) [reply pom.xml]: Resolved with the commit 2dd7e578286485ded51922fb59175ae57132af6c.

File: README.md (hunk 5 above)
Comment (Ketharan) [reply README.md]: Resolved with the commit 9c2f344de3cc07e51588e9804db3c8a37d8ae24b.

File: component/pom.xml (hunk 6 above)
Comment (Ketharan) [reply component/pom.xml]: Resolved with the commit d2b00308d9d9dd5d994adf46a1a1c8bbc0349edd.

File: component/pom.xml (hunk 7 above)
Comment (Ketharan) [reply component/pom.xml]: Resolved with the commit d2b0030.

File: component/src/main/java/org/wso2/extension/siddhi/io/nats/sink/NATSSink.java (hunk 8 above)
Comment (Ketharan) [reply component/src/main/java/org/wso2/extension/siddhi/io/nats/sink/NATSSink.java]: Resolved with the commit d16001bcee5871dcf30cfd65de771f93f3aeaaa1.

File: component/src/test/java/org/wso2/extension/siddhi/io/nats/sink/NATSSinkTestCase.java (hunk 9 above)
Comment (Ketharan) [reply component/src/test/java/org/wso2/extension/siddhi/io/nats/sink/NATSSinkTestCase.java]: Resolved with the commit 77f7e16ab934662a2382acbf160ccb0448d47bb4.

File: mkdocs.yml (hunk 10 above)
Comment (Ketharan) [reply mkdocs.yml]: Resolved with 765011d1d2b95fde0ecdf2215a58f6e98a1f91fa.

File: README.md (hunk 5 above)
Comment (Niveathika92) [reply README.md]: Features tag should be added by default in each build. See why that is not added

**Pull request merged**